File: day-02/solution.go
package main

import (
	cio "commons"
	"fmt"
	"strconv"
	"strings"
)

const (
	RED_MAX   = 12
	GREEN_MAX = 13
	BLUE_MAX  = 14
)

type Color string

var (
	RED   Color = "red"
	GREEN Color = "green"
	BLUE  Color = "blue"
)

type Game struct {
	red   int
	green int
	blue  int
}

type GameInfo struct {
	gameID int
	games  []Game
}

func main() {
	lines := cio.ReadLines("day-02/input.txt")
	gameIdSum := 0
	minPowerSum := 0
	for _, line := range lines {
		gameInfo := gameParser(line)
		allGamesValid := true
		minPowerSum += findPower(findMinimumSet(gameInfo))
		for _, game := range gameInfo.games {
			if !gameValidator(game) {
				allGamesValid = false
				break
			}
		}
		if allGamesValid {
			gameIdSum += gameInfo.gameID
		}
	}

	fmt.Println("Game ID Sum: ", gameIdSum)
	fmt.Println("Minimum Power Sum: ", minPowerSum)
}

func gameParser(line string) GameInfo {
	data := strings.Split(line, ":")
	gameID, _ := strconv.Atoi(strings.Split(data[0], " ")[1])
	games := []Game{}
	gameDatas := strings.Split(data[1], ";")
	for _, gameData := range gameDatas {
		colorData := strings.Split(gameData, ",")
		game := Game{}
		for _, colorData := range colorData {
			colorData = strings.TrimSpace(colorData)
			colorNum := strings.Split(colorData, " ")
			num, _ := strconv.Atoi(colorNum[0])
			color := Color(colorNum[1])
			switch color {
			case RED:
				game.red = num
			case GREEN:
				game.green = num
			case BLUE:
				game.blue = num
			}
		}
		games = append(games, game)
	}
	return GameInfo{
		gameID,
		games,
	}
}

func gameValidator(game Game) bool {
	if game.red > RED_MAX || game.green > GREEN_MAX || game.blue > BLUE_MAX {
		return false
	}
	return true
}

func findMinimumSet(gameInfo GameInfo) Game {
	games := gameInfo.games
	minimumSet := Game{}
	for _, game := range games {
		if game.red > minimumSet.red {
			minimumSet.red = game.red
		}
		if game.green > minimumSet.green {
			minimumSet.green = game.green
		}
		if game.blue > minimumSet.blue {
			minimumSet.blue = game.blue
		}
	}
	return minimumSet
}

func findPower(game Game) int {
	return game.red * game.green * game.blue
}

File: day-02/solution.rs
// cargo-deps: commons = { path = "../commons" }

const RED_MAX: i32 = 12;
const GREEN_MAX: i32 = 13;
const BLUE_MAX: i32 = 14;

enum Color {
    Red,
    Green,
    Blue,
}

struct Game {
    red: i32,
    green: i32,
    blue: i32,
}

struct GameInfo {
    game_id: i32,
    games: Vec<Game>,
}

fn main() {
    let lines = io::read_lines("day-02/input.txt");
    let mut min_power_sum = 0;
    let mut game_id_sum = 0;
    for line in lines {
        let game_info = game_parser(&line);
        let mut all_games_valid = true;
        min_power_sum += find_power(find_minimum_set(&game_info));
        for game in game_info.games {
            if !game_validator(game) {
                all_games_valid = false;
                break;
            }
        }
        if all_games_valid {
            game_id_sum += game_info.game_id;
        }
    }

    println!("Game ID Sum: {}", game_id_sum);
    println!("Minimum Power Sum: {}", min_power_sum);
}

fn game_parser(line: &str) -> GameInfo {
    let data: Vec<&str> = line.split(":").collect();
    let game_id = data[0].split(" ").collect::<Vec<&str>>()[1]
        .parse::<i32>()
        .unwrap();
    let mut games = Vec::new();
    let game_datas = data[1].split(";");
    for game_data in game_datas {
        let mut game = Game {
            red: 0,
            green: 0,
            blue: 0,
        };
        let color_data = game_data.split(",");
        for color_data in color_data {
            let color_data = color_data.trim();
            let color_num: Vec<&str> = color_data.split(" ").collect();
            let num = color_num[0].parse::<i32>().unwrap();
            let color = match color_num[1] {
                "red" => Color::Red,
                "green" => Color::Green,
                "blue" => Color::Blue,
                _ => panic!("Invalid color!"),
            };
            match color {
                Color::Red => game.red = num,
                Color::Green => game.green = num,
                Color::Blue => game.blue = num,
            };
        }
        games.push(game);
    }
    GameInfo { game_id, games }
}

fn game_validator(game: Game) -> bool {
    if game.red > RED_MAX || game.green > GREEN_MAX || game.blue > BLUE_MAX {
        return false;
    }
    true
}

fn find_minimum_set(game_info: &GameInfo) -> Game {
    let games = &game_info.games;
    let mut minimum_set = Game {
        red: 0,
        green: 0,
        blue: 0,
    };
    for game in games {
        if game.red > minimum_set.red {
            minimum_set.red = game.red;
        }
        if game.green > minimum_set.green {
            minimum_set.green = game.green;
        }
        if game.blue > minimum_set.blue {
            minimum_set.blue = game.blue;
        }
    }
    minimum_set
}

fn find_power(game: Game) -> i32 {
    game.red * game.green * game.blue
}